README.md
# JSON to envdir tool

In goes json, out goes envdir.

## JSON Format

```json
{
  "name": "myapp",
  "env": {
    "MYVAR": "123",
    "MYVAR_TMPL": "{{.UUID}} 123",
    "MYVAR_HEX": "{{.HEX 32}} 123",
    ...
  }
}
```

Notice that you can have dynamic value for an env var. The format is in Go template. Currently `{{.UUID}}` and `{{.HEX N}}` are supported. HEX accepts an integer for length - the resulting string will be double this length.

## Config File Format

**Default Config File**: `/etc/json2envdir.conf`

```ini
[envdir "webapp"]
path = /www/webapp/env.d

[envdir "myapp"]
path = /etc/myapp1/env.d
path = /etc/myapp2/env.d
file-perms = 0640
path-perms = 0750
```

## Usage

```bash
json2envdir --file=app-env.json
```


debian/changelog
json2envdir (3.0.0) focal; urgency=medium

  * Merge pull request #6 from heroku/amerine-do-not-panic-when-no-config
  * fix: Don't panic on unknown names.

 -- Mark Turner <mark.turner@salesforce.com> Wed, 13 Oct 2021 18:51:10 +0200

json2envdir (2.0.1) focal; urgency=medium

  * fix config path

 -- Ricardo Chimal Jr. <rchimal@salesforce.com>  Wed, 13 Oct 2021 14:27:23 +0200


json2envdir_test.go
package json2envdir

import (
	"io/ioutil"
	"os"
	"path/filepath"
	"testing"

	"github.com/heroku/json2envdir/config"
	"github.com/satori/go.uuid"
)

func TestParseNoConfig(t *testing.T) {
	testDir, err := ioutil.TempDir("", "json2envdirTest")
	if err != nil {
		t.Fatal(err)
	}
	defer func() { _ = os.RemoveAll(testDir) }()

	cfg := config.Config{
		Sections: map[string]*config.EnvDirSection{
			"cheetah": &config.EnvDirSection{
				Path: []string{testDir},
			},
		},
	}

	err = Process(cfg, `{
		"name": "not-cheetah",
		"env": {
			"VAR": "{{.UUID}}"
		}
	}`)

	if err != nil {
		t.Fatalf("Unexpected failure: %q", err)
	}
}

func TestParse(t *testing.T) {
	dir1, err := ioutil.TempDir("", "json2envdir1")
	if err != nil {
		t.Fatal(err)
	}
	defer os.RemoveAll(dir1)

	dir2, err := ioutil.TempDir("", "json2envdir2")
	if err != nil {
		t.Fatal(err)
	}
	defer os.RemoveAll(dir2)

	cfg := config.Config{
		Sections: map[string]*config.EnvDirSection{
			"myapp": &config.EnvDirSection{
				Path: []string{dir1, dir2},
			},
		},
	}
	err = Process(cfg, `{
		"name": "myapp",
		"env": {
			"MYVAR": "123",
			"MYVAR_TMPL": "{{.UUID}}",
			"MYVAR_HEX": "{{.HEX 32}}"
		}
	}`)
	if err != nil {
		t.Fatalf("Unexpected error in Parse: %s", err)
	}

	b, err := ioutil.ReadFile(filepath.Join(dir1, "MYVAR"))
	if err != nil {
		t.Fatal(err)
	}
	if string(b) != "123" {
		t.Fatalf("Value of $MYVAR in dir1 should equal to 123, but it's %s", b)
	}

	b, err = ioutil.ReadFile(filepath.Join(dir1, "MYVAR_TMPL"))
	if err != nil {
		t.Fatal(err)
	}
	uuid1, err := uuid.FromString(string(b))
	if err != nil {
		t.Fatalf("Value of $MYVAR_TMPL in dir1 should contain a UUID")
	}

	hex1, err := ioutil.ReadFile(filepath.Join(dir1, "MYVAR_HEX"))
	if err != nil {
		t.Fatal(err)
	}
	if len(hex1) != 64 {
		t.Fatalf("Length of $MYVAR_HEX in dir1 should be 64, got: %d", len(hex1))
	}

	b, err = ioutil.ReadFile(filepath.Join(dir2, "MYVAR"))
	if err != nil {
		t.Fatal(err)
	}
	if string(b) != "123" {
		t.Fatalf("Value of $MYVAR in dir2 should equal to 123, but it's %s", b)
	}

	b, err = ioutil.ReadFile(filepath.Join(dir2, "MYVAR_TMPL"))
	if err != nil {
		t.Fatal(err)
	}
	uuid2, err := uuid.FromString(string(b))
	if err != nil {
		t.Fatalf("Value of $MYVAR_TMPL in dir2 should contain a UUID")
	}

	hex2, err := ioutil.ReadFile(filepath.Join(dir2, "MYVAR_HEX"))
	if err != nil {
		t.Fatal(err)
	}
	if len(hex2) != 64 {
		t.Fatalf("Length of $MYVAR_HEX in dir1 should be 64, got: %d", len(hex1))
	}

	if uuid1 != uuid2 {
		t.Fatalf("uuid1 should equal to uuid2: uuid1=%s uuid2=%s", uuid1, uuid2)
	}
}


config/config_test.go
package config

import (
	"io/ioutil"
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

func TestLoadConfig(t *testing.T) {
	cfgStr := `[envdir "myapp"]
path = path1
path = path2
file-perms = 0640
path-perms = 0750
`

	dir, err := ioutil.TempDir("", "config")
	if err != nil {
		t.Fatal(err)
	}
	defer os.RemoveAll(dir)

	path := filepath.Join(dir, "example.conf")
	err = ioutil.WriteFile(path, []byte(cfgStr), os.ModePerm)
	if err != nil {
		t.Fatal(err)
	}

	cfg := LoadConfig(path)
	section := cfg.Sections["myapp"]
	if section == nil {
		t.Fatalf("no myapp section is found")
	}

	if !reflect.DeepEqual(section.Path, []string{"path1", "path2"}) {
		t.Fatalf("section should contain path1 and path2: %v", section.Path)
	}

	if section.PathPermsString != "0750" {
		t.Fatalf("path perms should equal to 0750", section.PathPermsString)
	}

	if section.FilePermsString != "0640" {
		t.Fatalf("file perms should equal to 0640", section.FilePermsString)
	}
}


json2envdir.go
package json2envdir

import (
	"bytes"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"html/template"
	"io/ioutil"
	"os"
	"path/filepath"

	"github.com/heroku/json2envdir/config"
	"github.com/satori/go.uuid"
)

type payload struct {
	Name string                 `json:"name"`
	Env  map[string]interface{} `json:"env"`
}

// Funcs is a set of functions that can be used in
// the template value of an env var
type Funcs struct {
}

// UUID is a function exposed to the templates for generating a UUIDv4
func (f Funcs) UUID() string {
	u := uuid.NewV4()
	return u.String()
}

// Hex allows templates to generate a hex of a specific length
func (f Funcs) HEX(length int) string {
	randomBytes := make([]byte, length)
	rand.Read(randomBytes)
	return hex.EncodeToString(randomBytes)
}

// Process parses and populates the envdirs specified in the config if matching
// data is found in the json payload.
func Process(cfg config.Config, rawJSON string) error {
	var p payload
	err := json.Unmarshal([]byte(rawJSON), &p)
	if err != nil {
		return err
	}

	envCfg, found := cfg.GetEnv(p.Name)
	if !found {
		fmt.Printf("Skipping unconfigured entry: %s", p.Name)
		return nil
	}

	funcs := Funcs{}
	for key := range p.Env {
		value := fmt.Sprintf("%v", p.Env[key])

		var result bytes.Buffer
		tmpl, err := template.New("").Parse(value)
		if err != nil {
			return err
		}
		err = tmpl.Execute(&result, funcs)
		if err != nil {
			return err
		}

		for _, path := range envCfg.Path {
			err = os.MkdirAll(path, envCfg.PathPerms)
			if err != nil {
				return err
			}

			err = ioutil.WriteFile(filepath.Join(path, key), result.Bytes(), envCfg.PathPerms)
			if err != nil {
				return err
			}
		}

	}
	return nil
}


cmd/json2envdir/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"

	"github.com/heroku/json2envdir"
	"github.com/heroku/json2envdir/config"
	flag "github.com/ogier/pflag"
)

var (
	jsonFile    = flag.StringP("file", "f", "-", "file with JSON environment")
	configFile  = flag.StringP("config", "c", config.DefaultConfigFile, "config file")
	versionFlag = flag.BoolP("version", "v", false, "show version")
	version     = "dev"
)

func readStdin() string {
	bytes, err := ioutil.ReadAll(os.Stdin)
	if err != nil {
		panic(err)
	}
	return string(bytes)
}

func readFile(filename string) string {
	bytes, err := ioutil.ReadFile(filename)
	if err != nil {
		panic(err)
	}
	return string(bytes)
}

func main() {
	flag.Parse()

	if *versionFlag {
		fmt.Println(version)
		return
	}

	cfg := config.LoadConfig(*configFile)

	var err error
	if *jsonFile == "-" {
		err = json2envdir.Process(cfg, readStdin())
	} else {
		err = json2envdir.Process(cfg, readFile(*jsonFile))
	}

	if err != nil {
		log.Fatal(err)
	}
}


config/config.go
package config

import (
	"os"
	"strconv"

	"gopkg.in/gcfg.v1"
)

var (
	// DefaultConfigFile is the default place json2envdir will look if
	// no config is passed in via a cli flag.
	DefaultConfigFile = "/etc/json2envdir.conf"

	defaultFilePerms os.FileMode = 0640
	defaultPathPerms os.FileMode = 0750
)

// Config holds parsed config.
type Config struct {
	Sections map[string]*EnvDirSection `gcfg:"envdir"`
}

// EnvDirSection represents a configured envdir diretive in the config.
type EnvDirSection struct {
	Path            []string
	PathPermsString string `gcfg:"path-perms"`
	FilePermsString string `gcfg:"file-perms"`

	PathPerms os.FileMode
	FilePerms os.FileMode
}

// LoadConfig loads the specified config file. If an error is encountered the program will panic.
func LoadConfig(configFile string) Config {
	if len(configFile) < 1 {
		configFile = DefaultConfigFile
	}

	cfg := Config{}
	err := gcfg.ReadFileInto(&cfg, configFile)
	if err != nil {
		panic(err)
	}
	return cfg
}

// GetEnv attempts to return the requested EnvDirSection and a boolean representing if the section was found.
// The caller should verify the section was found.
func (cfg Config) GetEnv(env string) (*EnvDirSection, bool) {
	if conf, ok := cfg.Sections[env]; ok {
		conf.parsePerms()
		return conf, true
	}

	return nil, false
}

func parsePerms(perms string, def os.FileMode) os.FileMode {
	if len(perms) > 0 {
		p, err := strconv.ParseUint(perms, 8, 32)
		if err != nil {
			panic(err)
		}
		return os.FileMode(p)
	}
	return def
}

func (e *EnvDirSection) parsePerms() {
	e.PathPerms = parsePerms(e.PathPermsString, defaultPathPerms)
	e.FilePerms = parsePerms(e.FilePermsString, defaultFilePerms)
}
